### PR TITLE
 Incorporating PIC state machine level retry changes into webrtc signaling state machine

### PR DESCRIPTION
**Pull request by kateyanurag**
…ling state machine

*Issue #, if available:*

*Description of changes:*
 Incorporating PIC state machine level retry changes into webrtc signaling state machine

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by disa6302 — COMMENTED** [review]
Based on discussion:

There are 2 levels of retries:
1. Application level backoff for `createSignalingClientSync()`
2. Internal PIC level retry for state transitions based on failure (`PSignalingClientRetryStrategy`)

We want to have a single `createSignalingClientSync()` which handles application level retry and internal PIC state machinery retry as well. In order to consolidate this, we will have the backoff state structure as part of clientInfo, which will be set up in the application. However, we want to get rid of  `createSignalingClientSyncWithRetryStrategy()` and let PIC level retries to happen internally. Customers will invoke `createSignalingClientSync()` only once.

File: src/source/Signaling/Client.c [hunk 1]
```diff
@@ -1,13 +1,37 @@
...
+STATUS validateSignalingClientRetryStrategy(PSignalingClientInfo pClientInfo) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSignalingClientRetryStrategy pSignalingClientRetryStrategy;
+
+    CHK(pClientInfo != NULL, STATUS_NULL_ARG);
+
+    pSignalingClientRetryStrategy = &(pClientInfo->signalingClientRetryStrategy);
+
+    CHK(pSignalingClientRetryStrategy->retryStrategyType > KVS_RETRY_STRATEGY_DISABLED &&
```
Comment (disa6302) [inline src/source/Signaling/Client.c:13]: `pSignalingClientRetryStrategy->retryStrategyType > KVS_RETRY_STRATEGY_DISABLED` to return `STATUS_NULL_ARG` seems misleading. Maybe we can look into a more suitable status code? What is the behavior in PIC if this retry strategy is used? I am wondering if `STATUS_NOT_IMPLEMENTED` would be better.

File: src/source/Signaling/Client.c [hunk 2]
```diff
@@ -16,7 +40,28 @@ STATUS createSignalingClientSync(PSignalingClientInfo pClientInfo, PChannelInfo
     MEMSET(&signalingClientInfoInternal, 0x00, SIZEOF(signalingClientInfoInternal));
     signalingClientInfoInternal.signalingClientInfo = *pClientInfo;
 
-    CHK_STATUS(createSignalingSync(&signalingClientInfoInternal, pChannelInfo, pCallbacks, pCredentialProvider, &pSignalingClient));
+    CHK_STATUS(validateSignalingClientRetryStrategy(pClientInfo));
```
Comment (disa6302) [inline src/source/Signaling/Client.c:43]: Same here. Do we want to pass the entire struct to validate retry strategy? No harm in passing it though.

**Review by disa6302 — COMMENTED** [review]
Overall logic looks good to me. I have added some minor comments.

File: samples/Common.c [hunk 3]
```diff
@@ -782,6 +803,8 @@ STATUS createSampleConfiguration(PCHAR channelName, SIGNALING_CHANNEL_ROLE_TYPE
     pSampleConfiguration->clientInfo.version = SIGNALING_CLIENT_INFO_CURRENT_VERSION;
     pSampleConfiguration->clientInfo.loggingLevel = logLevel;
     pSampleConfiguration->clientInfo.cacheFilePath = NULL; // Use the default path
+    CHK_STATUS(setupDefaultSignalingClientRetryStrategy(&pSampleConfiguration->clientInfo));
```
[the remark this answers is not in the document]
Comment (kateyanurag) [reply samples/Common.c:806]: Yes we need to pass entire client info.

Explanation:
I missed one change in common.c. I've added a new line - 
`pSignalingClientInfo->signalingClientCreationMaxRetryCount = MAX_CREATE_SIGNALING_CLIENT_RETRIES;`

This is for the outer retry while loop in createSignalingClientSync API. This parameter is defined in SignalingClientInfo struct. This parameter cannot go into SignalingClientRetryStrategy. SignalingClientRetryStrategy is basically KvsRetryStrategy defined in PIC.

```
typedef struct {
    UINT32 version;                                                
    CHAR clientId[MAX_SIGNALING_CLIENT_ID_LEN + 1];                
    UINT32 loggingLevel;                                  
    PCHAR cacheFilePath;             
    SignalingClientRetryStrategy signalingClientRetryStrategy;      
    UINT32 signalingClientCreationMaxRetryCount;              
} SignalingClientInfo, *PSignalingClientInfo;
```

File: src/source/Signaling/Client.c (hunk 1 above)
Comment (kateyanurag) [reply src/source/Signaling/Client.c:13]: Agree. STATUS_NOT_IMPLEMENTED would be better. Fixing it. I did not find that earlier. If KVS_RETRY_STRATEGY_DISABLED is used, then no wait times will be injected on failures (current behavior).

File: src/source/Signaling/Client.c (hunk 2 above)
Comment (kateyanurag) [reply src/source/Signaling/Client.c:43]: Yeah same reason. Need to make sure `pClientInfo->signalingClientCreationMaxRetryCount` is at least 1.

**Review by disa6302 — APPROVED** [review]

**Pull request merged**